README.md
# rk.CohenKappa

This R package provides a GUI to calculate Cohen's Kappa.

![rk.CohenKappa main tab](./inst/screenshots/main_tab.png?raw=true "rk.CohenKappa main tab")

It is a plugin package for [RKWard](https://rkward.kde.org), a powerful GUI and IDE for [R](https://r-project.org),
and will **only work in conjunction with RKWard**. However, RKWard is free software, please
[check it out](https://rkward.kde.org).

## Installation

### Installation via RKWard

After you installed RKWard, simply install this package like any other R package. The easiest way to get the most
recent stable release is to use RKWard's package management:

- Go to `Settings` -> `Manage R packages and plugins...` -> `Install/Update/Remove R packages`
- Check `Show only packages providing RKWard dialogs` and look at `New Packages`
- Check the desired package(s) and hit `Apply`

After you installed the package, RKWard should immediately have additional menu entries -- there's
usually no need to restart the application.

### Installation via GitHub

To install it directly from GitHub, you can use `install_github()` from the [devtools](https://github.com/hadley/devtools) package:

```
devtools::install_github("rkward-community/rk.CohenKappa") # stable release
devtools::install_github("rkward-community/rk.CohenKappa", ref="develop") # development release
```

You can safely ignore warnings about the (unavailable) package "rkward", it was installed together with RKWard.
 
## Contributing

See https://docs.kde.org/trunk5/en/rkward/rkwardplugins/index.html for documentation on writing plugins for RKWard.
Please contact the [RKWard development mailing list](https://mail.kde.org/mailman/listinfo/rkward-devel)
for help, reports and requests.

- Submit pull requests or patches
- Contact us to get write access

This plugin is built using the [rkwarddev](https://files.kde.org/rkward/R/pckg/rkwarddev/index.html) package. This means that its development is mainly focussed on one
script file, located at `rk.CohenKappa/inst/rkward/rkwarddev_plugin_script_kappa.R`. This script is written in `R` code and *generates*
all of the other plugin files when run. Therefore, in case you'd like to add to this plugin, please add to this script file.

### Branches

Please note that all development happens in the `develop` branch. Pull requests against the `master` branch will be rejected, as it is reserved for the current stable release.

## Licence

Copyright 2012-2022 Meik Michalke <meik.michalke@hhu.de>

rk.CohenKappa is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

rk.CohenKappa is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with rk.CohenKappa.  If not, see <http://www.gnu.org/licenses/>.


ChangeLog
ChangeLog for package rk.CohenKappa

changes in version 0.2-1 (2022-05-03)
unreleased:
  - this release is under development

changes in version 0.01-9 (2022-05-03)
fixed:
  - screenshot URL and removed flattr-link in README.md

changes in version 0.01-8 (2015-12-18)
added:
  - plugin is now translatable (see inst/rkward/po/)
  - german translation
changed:
  - updated the generator script
  - rebuilt with rkwarddev 0.08-1

changes in version 0.01-7 (2015-03-14)
fixed:
  - replaced the outdated links to sourceforge with our new KDE URL

changes in version 0.01-6 (2015-03-13)
changed:
  - rebuilt using rkwarddev 0.07-1

changes in version 0.01-5 (2013-04-04)
added:
  - debianized the plugin
fixed:
  - removed erroneous pluginmap inclusions from XML files

changes in version 0.01-4 (2013-02-24)
changed:
  - updated rkwarddev generator script (compatibility with 0.06-2) and regenerated plugin code

changes in version 0.01-3 (2012-04-07)
changes:
  - added a ChangeLog file
  - rebuild package for R 2.15


inst/rkward/plugins/CohensKappa.js
// this code was generated using the rkwarddev package.
// perhaps don't make changes here, but in the rkwarddev script instead!



function preprocess(is_preview){
  // add requirements etc. here
  echo("require(psych)\n");
}

function calculate(is_preview){
  // read in variables from dialog
  var data = getString("data");
  var dataWeight = getString("dataWeight");
  var nobs = getString("nobs");
  var alpha = getString("alpha");

  // the R code to be evaluated
  echo("  kappa.result <- cohen.kappa(");
  if(data) {
    echo("\n    x=" + data);  
  } else {}
  if(dataWeight) {
    echo(",\n    w=" + dataWeight);  
  } else {}
  if(nobs > 0) {
    echo(",\n    n.obs=" + nobs);  
  } else {}
  if(alpha != 0.05) {
    echo(",\n    alpha=" + alpha);  
  } else {}
  echo("\n  )\n\n");
}

function printout(is_preview){
  // printout the results
  new Header(i18n("Cohen's Kappa and weighted Kappa")).print();
  new Header(i18n("Correlation coefficients and confidence boundaries"), 3).add(i18n("Alpha level"), getValue("alpha")).add(i18n("Number of subjects"), noquote("kappa.result[[\"n.obs\"]]")).print();
  echo("rk.print(kappa.result[[\"confid\"]])\n");
  //// save result object
  // read in saveobject variables
  var saveResults = getValue("saveResults");
  var saveResultsActive = getValue("saveResults.active");
  var saveResultsParent = getValue("saveResults.parent");
  // assign object to chosen environment
  if(saveResultsActive) {
    echo(".GlobalEnv$" + saveResults + " <- kappa.result\n");
  } else {}

}



inst/rkward/rkwarddev_plugin_script_kappa.R
# the plugin code was generated by this script
# you should not change the plugin code directly, but this script
# note: this script only creates objects in your workspace,
# *EXCEPT* for the last call, see below.

require(rkwarddev)
rkwarddev.required("0.08-1")

local({
# set the output directory to overwrite the actual plugin
output.dir <- tempdir()
overwrite <- TRUE
# if you set guess.getters to TRUE, the resulting code will need RKWard >= 0.6.0
guess.getter <- TRUE
rk.set.indent(by="  ")
rk.set.empty.e(TRUE)
update.translations <- TRUE

about.info <- rk.XML.about(
  name="rk.CohenKappa",
  author=c(
    person(given="Meik", family="Michalke",
      email="meik.michalke@hhu.de", role=c("aut","cre"))),
  about=list(desc="RKWard GUI to calculate Cohen's Kappa",
    version="0.2-1", url="https://rkward.kde.org", long.desc="RKWard GUI to calculate Cohen's Kappa, i.e., the agreement of two raters (using the psych package)")
  )
dependencies.info <- rk.XML.dependencies(
  dependencies=list(rkward.min=ifelse(isTRUE(guess.getter), "0.6.0", "0.5.6")),
  package=list(c(name="psych"))
)

############
## cohen's kappa
############
varSelect <- rk.XML.varselector(label="Select data", id.name="varSelect")
data <- rk.XML.varslot(label="Data (matrix or data.frame, 2*n or p*p)", source=varSelect, classes=c("data.frame", "matrix"), required=TRUE, id.name="data")
dataWeight <- rk.XML.varslot(label="Weight matrix (p*p)", source=varSelect, classes="matrix", id.name="dataWeight")

nobs <- rk.XML.spinbox(label="Number of observations (if data is a square matrix)", min=0, real=FALSE, id.name="nobs")
alpha <- rk.XML.spinbox(label="Alpha value for confidence interval", min=0, max=1, initial=0.05, id.name="alpha")

saveResults <- rk.XML.saveobj("Save results to workspace", initial="kappa.result", id.name="saveResults")

kappa.full.dialog <- rk.XML.dialog(
  rk.XML.row(
    varSelect,
    rk.XML.col(
      data,
      dataWeight,
      rk.XML.stretch(),
      rk.XML.frame(nobs,
      alpha),
      saveResults
    )
  ),
  label="Cohen's Kappa"
)

## JavaScript
kappa.js.calc <- rk.paste.JS(
  echo("  kappa.result <- cohen.kappa("),
  js(
    if(data){
      echo("\n    x=", data)
    } else {},
    if(dataWeight){
      echo(",\n    w=", dataWeight)
    } else {},
    if(nobs > 0){
      echo(",\n    n.obs=", nobs)
    } else {},
    if(alpha != 0.05){
      echo(",\n    alpha=", alpha)
    } else {}
  ),
  echo("\n  )\n\n")
)

kappa.js.print <- rk.paste.JS(
  rk.JS.header("Correlation coefficients and confidence boundaries",
    level=3,
    .add=list(
      add=c("Alpha level", alpha),
      add=c("Number of subjects", noquote="kappa.result[[\\\"n.obs\\\"]]")
    )
  ),
  echo("rk.print(kappa.result[[\"confid\"]])\n")
)


#############
## if you run the following function call, files will be written to tempdir!
#############
# this is where it get's serious, that is, here all of the above is put together into one plugin

rk.plugin.skeleton(
  about.info,
  path=output.dir,
  guess.getter=guess.getter,
  xml=list(
    dialog=kappa.full.dialog),
  js=list(results.header="Cohen's Kappa and weighted Kappa",
    require="psych",
     calculate=kappa.js.calc,
     printout=kappa.js.print),
  pluginmap=list(name="Cohen's Kappa", hierarchy=list("analysis")),
  dependencies=dependencies.info,
  create=c("pmap", "xml", "js", "desc"),
  overwrite=overwrite,
  tests=FALSE,
#  edit=TRUE,
  load=TRUE,
#  show=TRUE,
  hints=FALSE)

  if(isTRUE(update.translations)){
    rk.updatePluginMessages(file.path(output.dir,"rk.CohenKappa","inst","rkward","rk.CohenKappa.pluginmap"))
  } else {}
})
